### PR TITLE
set prefs

### PR DESCRIPTION
**Pull request by royreznik**
Marine-core pr:
https://github.com/tomlegkov/marine-core/issues/19

### CODE REVIEW

File: marine/marine.py [hunk 1]
```diff
@@ -10,10 +10,16 @@ class MarineResult(Structure):
 
 
 MARINE_RESULT_POINTER = POINTER(MarineResult)
+MARINE_PREF_T = c_char * 1024
```
Comment (yehonatanz) [inline marine/marine.py:13]: Expose it in `marine.h` and then get it from `libmarine.so`.

File: marine/marine.py [hunk 2]
```diff
@@ -105,6 +114,11 @@ def validate_display_filter(self, display_filter: str) -> bool:
         display_filter = display_filter.encode("utf-8")
         return bool(self._marine.validate_display_filter(display_filter))
 
+    def set_preferences(self, preferences: List[str]) -> int:
+        preferences = [MARINE_PREF_T(*pref.encode("utf-8")) for pref in preferences]
```
Comment (yehonatanz) [inline marine/marine.py:118]: For a moment there I thought you're pointer-dereferencing in python 😄

File: marine/marine.py [hunk 3]
```diff
@@ -105,6 +114,11 @@ def validate_display_filter(self, display_filter: str) -> bool:
...
+        arr = (MARINE_PREF_T * (len(preferences)))(*preferences)
+        return int(self._marine.set_preferences(arr, len(arr)))
```
Comment (yehonatanz) [inline marine/marine.py:120]: That's python, not C.
Raise an exception instead of setting an error code.

File: marine/marine.py [hunk 4]
```diff
@@ -10,10 +10,16 @@ class MarineResult(Structure):
...
+MARINE_PREF_T = c_char * 1024
 
 
 class Marine:
-    def __init__(self, lib_path: str, epan_auto_reset_count: Optional[int] = None):
+    def __init__(
+        self,
+        lib_path: str,
+        epan_auto_reset_count: Optional[int] = None,
+        preferences: List[str] = None,
```
Comment (yehonatanz) [inline marine/marine.py:21]: `Optional[List[str]]`

File: marine/marine_pool.py [hunk 5]
```diff
@@ -44,9 +43,29 @@ def filter_and_parse(
             chunksize=chunk_size,
         )
 
+    def set_preferences(self, preferences: List[str]) -> int:
+        # Because each process have a marine instance, and there is no way to tell each process to run
+        # the _marine_instance.set_preferences, we need to recreate the pool with the new configuration
+        self.preferences = preferences
+        self.pool.close()
```
Comment (yehonatanz) [inline marine/marine_pool.py:50]: That's disgusting.
What's wrong with `self.pool.apply(cls._set_preferences, preferences)`?

**Review by yehonatanz — CHANGES_REQUESTED** [review]

File: marine/marine.py (hunk 1 above)
Comment (tomlegkov) [reply marine/marine.py:13]: I'm not a big fan of this approach. Why not have the API receive a simple string and convert it to `marine_pref_t` inside the C code?

File: marine/marine.py [hunk 6]
```diff
@@ -105,6 +114,11 @@ def validate_display_filter(self, display_filter: str) -> bool:
         display_filter = display_filter.encode("utf-8")
         return bool(self._marine.validate_display_filter(display_filter))
 
+    def set_preferences(self, preferences: List[str]) -> int:
```
Comment (tomlegkov) [inline marine/marine.py:117]: How complicated would it be to add a getter function to this? If it takes more than a few minutes it's not worth it, but otherwise I think it's a nice feature.

File: marine/marine_pool.py [hunk 7]
```diff
@@ -10,21 +10,20 @@ class MarinePool:
...
+        lib_path: str,
+        epan_auto_reset_count: int,
+        process_count: int = 4,
+        preferences: Optional[List[str]] = None,
     ):
+        self.pool = None
         self._lib_path = lib_path
         self._epan_auto_reset_count = epan_auto_reset_count
         self._process_count = process_count
+        self.preferences = preferences
```
Comment (tomlegkov) [inline marine/marine_pool.py:23]: Why is this not private like the others?

File: tests/marine/test_marine.py [hunk 8]
```diff
@@ -49,8 +49,8 @@ def general_filter_and_parse_test(
         {k: str(v) for k, v in expected_output.items()} if expected_output else None
     )
 
-    assert expected_passed == passed
-    assert expected_output == output
+    assert passed == expected_passed
```
Comment (tomlegkov) [inline tests/marine/test_marine.py:52]: Isn't the convention `expected == actual`? Why did you change it?

File: tests/marine/test_marine.py [hunk 9]
```diff
@@ -326,6 +326,43 @@ def test_filter_and_parse_without_fields(
     )
 
 
+def test_preferences_change_effect_while_running(
```
Comment (tomlegkov) [inline tests/marine/test_marine.py:329]: I really like this test, nice!

If you decide to add the getter function to prefs, also assert that the tcp checksum verification is actually 0.

**Review by tomlegkov — CHANGES_REQUESTED** [review]

File: marine/marine_pool.py (hunk 7 above)
Comment (royreznik) [reply marine/marine_pool.py:23]: why should it? its a nice way to get the perefenrces here

File: marine/marine_pool.py (hunk 5 above)
Comment (royreznik) [reply marine/marine_pool.py:50]: You can't

File: tests/marine/test_marine.py (hunk 8 above)
Comment (royreznik) [reply tests/marine/test_marine.py:52]: yea i was confused too, but pycharm tells me its `acutal == excpetion`

File: marine/marine.py (hunk 1 above)
Comment (royreznik) [reply marine/marine.py:13]: i thought about this, i just didn't want to add those malloc, strcpy and free in the `C` code

File: marine/marine.py (hunk 6 above)
Comment (royreznik) [reply marine/marine.py:117]: it's can be easily done in the python code, but its gonna be only the pref's you set from the python

if i want to add this in the C i think it's gonna take a while, cause i need to save all the preferences you set in array or something, and i think its not worth it, i'm guessing also its gonna be more useful in the python code (because noeone really use the C API without the python wrap)

File: marine/marine.py (hunk 2 above)
Comment (royreznik) [reply marine/marine.py:118]: yes i can see that too 🤣

File: tests/marine/test_marine.py [hunk 10]
```diff
@@ -421,3 +458,7 @@ def test_set_epan_auto_reset_count(marine_instance: Marine):
     assert marine_instance.epan_auto_reset_count != SOME_VALUE
     marine_instance.epan_auto_reset_count = SOME_VALUE
     assert marine_instance.epan_auto_reset_count == SOME_VALUE
+
+
+def test_set_invalid_preference(marine_instance: Union[Marine]):
```
Comment (royreznik) [inline tests/marine/test_marine.py:463]: Well i add comment to my self: change type

File: marine/marine_pool.py (hunk 7 above)
Comment (tomlegkov) [reply marine/marine_pool.py:23]: What's the application of getting the preferences? I already know them since I passed them to the constructor

File: marine/marine.py (hunk 6 above)
Comment (tomlegkov) [reply marine/marine.py:117]: Wireshark probably has a getter function for prefs (since it can display them in the gui :)) without storing them yourself. If it's easy to use then consider adding it.

File: tests/marine/test_marine.py (hunk 8 above)
Comment (tomlegkov) [reply tests/marine/test_marine.py:52]: Use notepad++

**Comment by royreznik** [comment]
will we reopen soon
